### ISSUE TITLE
write script to send fedmsg when we want a release artifact signed

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#198]
As part of the [Fedora Signing Discussion](https://github.com/coreos/fedora-coreos-tracker/issues/187) it was identified that we need to send a message when we want to get our artifacts signed in preparation for a release. The payload of the message needs to include the artifacts as well as the checksums (so the signer can check before creating signature). We believe the `meta.json` should suffice as the payload.

This script would need to be executed somewhere in our release process, or perhaps any time a release candidate becomes available so that all signatures are in place well in time before we do a release.

**Comment by puiterwijk** [comment]
Note that fedmsg is being replaced with fedora-messaging.
I think you should use the latter for any new code.

**Comment by dustymabe** [comment]
> Note that fedmsg is being replaced with fedora-messaging.
> I think you should use the latter for any new code.

Indeed that is defnitely the plan. The problem I have when describing *fedora messaging messages* is that there isn't a succinct way to do it like there was with *fedmsg*. I've just still been calling them *fedmsgs* because people know what you mean when you say it.

**Comment by dustymabe** [comment]
After some recent discussions there are actually a few different messages we probably want to send. Here is a proposal for now:

- `org.fedoraproject.prod.coreos.ostree`
    - gives information about an ostree commit that was just created
    - can be used to trigger signing an ostree commit
    - similar to [pungi-compose-ostree](https://fedora-fedmsg.readthedocs.io/en/latest/topics.html#pungi-compose-ostree) message today.

- `org.fedoraproject.prod.coreos.build.status.change`
    - shares information about status of the current coreos build in jenkins (moving through stages of pipeline, etc)
    - similar to [pungi-compose-status-change](https://fedora-fedmsg.readthedocs.io/en/latest/topics.html#pungi-compose-status-change)
    - should deliver the information in the meta.json from a fedora coreos pipeline run

Any other ones I'm missing?

**Comment by jlebon** [comment]
> `org.fedoraproject.prod.coreos.ostree`

Maybe let's split this out into two, e.g. `org.fedoraproject.prod.coreos.ostree.signing-request` and `org.fedoraproject.prod.coreos.ostree`? The former is for #200; signing an OSTree commit during the build process so we can include it in the image artifacts. The latter would be when the OSTree commit is available in the bucket proper, and signalling to releng to pull it into the unirepo?

Actually, we could also fold the `signing-request` one into `org.fedoraproject.prod.coreos.build.status.change` messages. Not sure if that's harder for releng though.

**Comment by dustymabe** [comment]
> > `org.fedoraproject.prod.coreos.ostree`
> 
> Maybe let's split this out into two, e.g. `org.fedoraproject.prod.coreos.ostree.signing-request` and `org.fedoraproject.prod.coreos.ostree`? The former is for #200; signing an OSTree commit during the build process so we can include it in the image artifacts. The latter would be when the OSTree commit is available in the bucket proper, and signalling to releng to pull it into the unirepo?
> 
> Actually, we could also fold the `signing-request` one into `org.fedoraproject.prod.coreos.build.status.change` messages. Not sure if that's harder for releng though.

yeah maybe we should just do that for all of these and re-use the state change messages. Actually I just looked a bit closer and the pungi messages have a status change and a phase change. I guess a new proposal would be:

- `org.fedoraproject.prod.coreos.build.stage.start`
    - shares information about the COSA build moving through the pipeline 
    - similar to [pungi-compose-phase-start](https://fedora-fedmsg.readthedocs.io/en/latest/topics.html#pungi-compose-phase-start)
    - should deliver the information in the meta.json from a fedora coreos pipeline run
    - can be used to trigger ostree commit signing
    - can be used to trigger artifact signing
    - can be used to trigger transfer ostree to unified repo

- `org.fedoraproject.prod.coreos.build.status.change`
    - shares information about status of the current coreos COSA build in jenkins (i.e. whether it failed or not)
    - similar to [pungi-compose-status-change](https://fedora-fedmsg.readthedocs.io/en/latest/topics.html#pungi-compose-status-change)
    - should deliver the information in the meta.json from a fedora coreos pipeline run

**Comment by jlebon** [comment]
So `org.fedoraproject.prod.coreos.build.status.change` would only actually be emitted at the end of builds with the final success/failure result? Minor bikeshed, though ISTM like those should be called e.g. `org.fedoraproject.prod.coreos.build.complete` or something if that's the case.

Anyway, SGTM!

**Comment by dustymabe** [comment]
> So `org.fedoraproject.prod.coreos.build.status.change` would only actually be emitted at the end of builds with the final success/failure result?

Yeah when it starts and when it finishes (or fails). 

I'll take this to fedora infra and see if they have any input.

**Comment by dustymabe** [comment]
> I'll take this to fedora infra and see if they have any input.

https://pagure.io/fedora-infrastructure/issue/7940

**Comment by jlebon** [comment]
I have a suggestion: I think we should make more explicit the difference between informational messages ("if anyone cares, our build has reached state $x") from messages requesting direct action from releng ("please do $x for me; I'll be waiting").

For informational events we can start with:
- build started
- build ended

And we can add more messages in the future about intermediate states if useful.

For requests, there are three main ones:
- OSTree commit signing
- Build artifact signing
    - These two requests would ideally happen on every build (see https://github.com/coreos/fedora-coreos-tracker/issues/200#issuecomment-500054869). In theory, build artifact signing could happen as part of a release only, but signing every build I think would help with e2e testing.
- OSTree commit import into unified repo
    - This request will only happen during a release.

---

Strawman for informational events:

```
topic: org.fedoraproject.prod.coreos.build.status.change
headers:
    build_id: ${build_id}
    stream: ${stream}
    basearch: ${basearch}
    status: (STARTED|FINISHED|...)
message:
    # if FINISHED
    result: (SUCCESS|FAILURE)
    # if SUCCESS
    build_dir: s3://.../${build_id}/${basearch}/
```

Strawman for OSTree commit signing request:

```
topic: org.fedoraproject.prod.coreos.build.request.ostree-sign
headers:
    build_id: ${build_id}
    stream: ${stream}
    basearch: ${basearch}
message:
    commit_object: s3://.../${build_id}/${basearch}/ostree-commit-object
    checksum: <SHA256>
```

Strawman for build artifact signing request:

```
topic: org.fedoraproject.prod.coreos.build.request.artifacts-sign
headers:
    build_id: ${build_id}
    stream: ${stream}
    basearch: ${basearch}
message:
    metadata: s3://.../${build_id}/${basearch}/meta.json
    checksum: <SHA256>
```

Strawman for OSTree commit import requests:

```
topic: org.fedoraproject.prod.coreos.build.request.ostree-import
headers:
    build_id: ${build_id}
    stream: ${stream}
    basearch: ${basearch}
message:
    commit: s3://.../ostree-commit.tar
    # checksum for integrity checking the tarball; but note the commit inside will already be signed
    checksum: <SHA256>
```

And we'll need matching messages from Fedora infra to signal to us when the request was fulfilled, maybe just reflecting back the same message but on topic `$topic.finished` with header `status: (SUCCESS|FAILURE)`?

Does that look sane?
/cc @bgilbert @sinnykumari @puiterwijk

**Comment by bgilbert** [comment]
For `org.fedoraproject.prod.coreos.build.request.artifacts-sign`, do we want `release.json` rather than `meta.json`?  Or is that not physically possible?

LGTM otherwise.

**Comment by jlebon** [comment]
> For `org.fedoraproject.prod.coreos.build.request.artifacts-sign`, do we want `release.json` rather than `meta.json`? Or is that not physically possible?

Actually, I think it might be cleaner to just go the other and give the explicit list of artifacts we want signatures for. The advantage is that code on the infra side doesn't need to learn how to read `meta.json` or `release.json` and doesn't need to be changed when we add a new artifact type.

I amended my proposal above!

**Comment by abompard** [comment]
I'm not sure you actually need to use the message headers, I think those values would be more suited for the message body. Consumers of fedora-messaging messages don't usually look at message headers, unless they're doing something advanced or meta (like comparing messages, generating notifications, etc).

**Comment by jlebon** [comment]
> I'm not sure you actually need to use the message headers, I think those values would be more suited for the message body.

My understanding was that header fields can be selected on, so that consumers can be more specific in the event they're interested in. Is that correct? For example, when waiting for `org.fedoraproject.prod.coreos.build.request.artifacts-sign.finished` from releng, we can also select for the right `build_id`, rather than having to process the body itself to find out.

(My experience with the message bus mostly comes from fedmsg + https://github.com/jenkinsci/jms-messaging-plugin, but the internal broker that proxied fedmsgs had some magic which pulled out fields from the body and put them into headers so Jenkins jobs could trigger on them more accurately.)

> Consumers of fedora-messaging messages don't usually look at message headers, unless they're doing something advanced or meta

Gotcha. Don't want to go against the convention if that's what most fedora-messaging msgs do.

**Comment by sinnykumari** [comment]
Agree with what @abompard said, consumers generally uses content from msg section. So, all needed information which consumers may need should be in message section.

Few additional comment on proposal at https://github.com/coreos/fedora-coreos-tracker/issues/198#issuecomment-513944390 :
- To keep consistent with other messaging topic, let's use only dot(.) in topic name. Like rename topic  org.fedoraproject.prod.coreos.build.request.ostree-import to org.fedoraproject.prod.coreos.build.request.ostree.import
- I think we don't need  to send topic `org.fedoraproject.prod.coreos.build.request.ostree-sign`. Consumer can listen to topic `org.fedoraproject.prod.coreos.build.status.change` and sign ostree commit when they are in FINISHED state, otherwise do nothing.

**Comment by jlebon** [comment]
> Agree with what @abompard said, consumers generally uses content from msg section. So, all needed information which consumers may need should be in message section.

I can change this, I just want more confirmation about the implications of doing this as mentioned in https://github.com/coreos/fedora-coreos-tracker/issues/198#issuecomment-514226457.

> To keep consistent with other messaging topic, let's use only dot(.) in topic name

Sure, done!

> I think we don't need to send topic org.fedoraproject.prod.coreos.build.request.ostree-sign. Consumer can listen to topic org.fedoraproject.prod.coreos.build.status.change and sign ostree commit when they are in FINISHED state, otherwise do nothing.

OSTree commits need to be signed during the build, not at the end only, because they need to be included into the image artifacts (see discussions in #200). So we need to explicitly wait for the request to be fulfilled before we can go ahead with the rest of the build.

**Comment by sinnykumari** [comment]
> I can change this, I just want more confirmation about the implications of doing this as mentioned in [#198 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/198#issuecomment-514226457).

+1

> OSTree commits need to be signed during the build, not at the end only, because they need to be included into the image artifacts (see discussions in #200). So we need to explicitly wait for the request to be fulfilled before we can go ahead with the rest of the build.

sgtm

**Comment by abompard** [comment]
> My understanding was that header fields can be selected on, so that consumers can be more specific in the event they're interested in. Is that correct? For example, when waiting for `org.fedoraproject.prod.coreos.build.request.artifacts-sign.finished` from releng, we can also select for the right `build_id`, rather than having to process the body itself to find out.

The broker we use has that feature but we generally don't make use of it in our wrapper library, because it would make the binding phase much more complex. So I would recommend filtering only on the topic, and filtering further in the code itself.

**Comment by jlebon** [comment]
OK, updated to just fold headers into the message body!
(Also note that the actual bodies will be in JSON, not YAML as shown here).

**Comment by abompard** [comment]
Thanks!

**Comment by jlebon** [comment]
This is now blocked on https://pagure.io/fedora-infrastructure/issue/7940#comment-584414.

**Comment by dustymabe** [comment]
> Strawman for OSTree commit import requests:
> 
> ```
> topic: org.fedoraproject.prod.coreos.build.request.ostree-import
> message:
>    build_id: ${build_id}
>    stream: ${stream}
>    basearch: ${basearch}
>    commit: s3://.../ostree-commit.tar
>    # checksum for integrity checking the tarball; but note the commit inside will already be signed
>    checksum: <SHA256>
> ```

If we end up using this for [both the annex repo and the prod repo](https://github.com/coreos/fedora-coreos-tracker/issues/266#issuecomment-532352365) we should maybe add a field that specifies which repo we want to import into.

**Comment by jlebon** [comment]
> If we end up using this for both the annex repo and the prod repo we should maybe add a field that specifies which repo we want to import into.

So if we use the annex for storing all the OSTree streams, then at release time, should we instead have a different message for "`pull-local` this OSTree ref from the compose repo into prod" ?

**Comment by dustymabe** [comment]
> So if we use the annex for storing all the OSTree streams, then at release time, should we instead have a different message for "`pull-local` this OSTree ref from the compose repo into prod" ?

Yeah. Except I think we can just use the same message and just make the code:

- pull commit hash from fedora message
-  check the compose repo for the commit
- if exists in compose repo then pull-local from there
- if not then pull from s3

WDYT?

**Comment by jlebon** [comment]
OK, I added the following field to the `ostree-import` message:

```
    target_repo: <compose|prod>
```

Does that work?

I opted for symbolic "compose" and "prod" names rather than hurling internal Fedora infra URLs at itself. (Of course, this is all still conditional on whether we can use the compose repo for this, right? Or did you get an answer about this?)

**Comment by dustymabe** [comment]
> ```
>  target_repo: <compose|prod>
> ```
> 
> 
> Does that work?

yep. can we also add the commit hash ?



> I opted for symbolic "compose" and "prod" names rather than hurling internal Fedora infra URLs at itself. (Of course, this is all still conditional on whether we can use the compose repo for this, right? Or did you get an answer about this?)

I think that sounds good. I believe we can use the compose repo, but I'll get confirmation.

**Comment by dustymabe** [comment]
> > I opted for symbolic "compose" and "prod" names rather than hurling internal Fedora infra URLs at itself. (Of course, this is all still conditional on whether we can use the compose repo for this, right? Or did you get an answer about this?)
> 
> I think that sounds good. I believe we can use the compose repo, but I'll get confirmation.

I talked with @mohanboddu in IRC and he confirmed this should be OK:

```
14:46:03        dustymabe | nirik: mboddu quick question for you
14:46:24          +mboddu | dustymabe: shoot
14:46:35        dustymabe | we want to import our ostree builds from our composes into the compose ostree repo
                          | (the one pungi uses)
14:47:00        dustymabe | the idea is that we have two ostree repos today, and we'd like to just keep it at two
14:47:39        dustymabe | I just wanted to get explicit acknowledgement from you all that it was OK to do that
14:47:59        dustymabe | there will still be the prod repo where releases get synced to
14:48:51          +mboddu | two ostrees repo = prod repo + compose repo, right?
14:49:02        dustymabe | correct, no change from today
14:49:03          +mboddu | dustymabe: And we thought that we will keep both of them
14:49:25        dustymabe | yes we need to keep both of them because silverblue is still composing into the
                          | compose repo
14:49:38          +mboddu | Yes
14:50:04        dustymabe | mboddu: so you are OK with us importing the FCOS commits into the compose repo?
14:50:52          +mboddu | dustymabe: I guess so, but I am wondering if it at all affects silverblue somehow...?
14:51:17        dustymabe | mboddu: it should not, it will use different branch names, just like atomic host did
14:51:42          +mboddu | Right
14:53:52          +mboddu | But isn't fcos will mix and match different builds that might not be in stable?
14:54:24          +mboddu | Doesn't it affect the the repos?
14:54:44          +mboddu | s/repos/silverblue/
14:59:26        dustymabe | mboddu: the ostree repo is the place we store the already composed ostree commits. The
                          | inputs to composes come from yum repos. Silverblue will continue to pull from the
                          | locations it's currently pulling from
14:59:49        dustymabe | so this won't affect silverblue at all
15:00:03          +mboddu | dustymabe: Okay, then I am good with it
15:00:07        dustymabe | +1 thanks

```

**Comment by dustymabe** [comment]
in places where we have checksums let's qualify it by including the checksum algorithm:

```
checksum: <SHA256>
```

WDYT?

becomes

```
checksum: 'sha256:aabbccddee....'
```

**Comment by jlebon** [comment]
> in places where we have checksums let's qualify it by including the checksum algorithm:

Yeah, that makes sense. We'll have to update RoboSignatory too.

**Comment by jlebon** [comment]
Fixed by https://github.com/coreos/coreos-assembler/pull/763.
